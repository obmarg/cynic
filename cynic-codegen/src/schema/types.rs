use std::marker::PhantomData;

use super::type_index::TypeIndex;

#[derive(Debug, Clone, PartialEq)]
pub enum Type<'a> {
    Scalar(ScalarType<'a>),
    Object(ObjectType<'a>),
    Interface(InterfaceType<'a>),
    Union(UnionType<'a>),
    Enum(EnumType<'a>),
    InputObject(InputObjectType<'a>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum InputType<'a> {
    Scalar(ScalarType<'a>),
    Enum(EnumType<'a>),
    InputObject(InputObjectType<'a>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum OutputType<'a> {
    Scalar(ScalarType<'a>),
    Object(ObjectType<'a>),
    Interface(InterfaceType<'a>),
    Union(UnionType<'a>),
    Enum(EnumType<'a>),
}

#[derive(Debug, Clone, PartialEq)]
pub struct ScalarType<'a> {
    pub name: &'a str,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ObjectType<'a> {
    pub description: Option<&'a str>,
    pub name: &'a str,
    // pub implements_interfaces: Vec<InterfaceRef>,
    pub fields: Vec<Field<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Field<'a> {
    pub description: Option<&'a str>,
    pub name: &'a str,
    pub arguments: Vec<InputValue<'a>>,
    pub field_type: TypeRef<'a, OutputType<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct InputValue<'a> {
    pub description: Option<&'a str>,
    pub name: &'a str,
    pub value_type: TypeRef<'a, InputType<'a>>,
    // pub default_value: Option<Value<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct InterfaceType<'a> {
    pub description: Option<&'a str>,
    pub name: &'a str,
    pub fields: Vec<Field<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct UnionType<'a> {
    pub description: Option<&'a str>,
    pub name: &'a str,
    pub types: Vec<&'a str>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct EnumType<'a> {
    pub description: Option<&'a str>,
    pub name: &'a str,
    pub values: Vec<EnumValue<'a>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct EnumValue<'a> {
    pub description: Option<&'a str>,
    pub name: &'a str,
}

#[derive(Debug, Clone, PartialEq)]
pub struct InputObjectType<'a> {
    pub description: Option<&'a str>,
    pub name: &'a str,
    pub fields: Vec<InputValue<'a>>,
}

#[derive(Clone)]
pub enum TypeRef<'a, T> {
    Named(&'a str, &'a TypeIndex<'a>, PhantomData<fn() -> T>),
    List(Box<TypeRef<'a, T>>),
    Nullable(Box<TypeRef<'a, T>>),
}

impl<'a, T> PartialEq for TypeRef<'a, T> {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Self::Named(l0, l1, _), Self::Named(r0, r1, _)) => l0 == r0 && std::ptr::eq(l1, r1),
            (Self::List(l0), Self::List(r0)) => l0 == r0,
            (Self::Nullable(l0), Self::Nullable(r0)) => l0 == r0,
            _ => false,
        }
    }
}

impl std::fmt::Debug for TypeRef<'_, InputType<'_>> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Named(arg0, _, _) => f.debug_tuple("NamedInputType").field(arg0).finish(),
            Self::List(arg0) => f.debug_tuple("ListType").field(arg0).finish(),
            Self::Nullable(arg0) => f.debug_tuple("NullableType").field(arg0).finish(),
        }
    }
}

impl std::fmt::Debug for TypeRef<'_, OutputType<'_>> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Named(arg0, _, _) => f.debug_tuple("NamedOutputType").field(arg0).finish(),
            Self::List(arg0) => f.debug_tuple("ListType").field(arg0).finish(),
            Self::Nullable(arg0) => f.debug_tuple("NullableType").field(arg0).finish(),
        }
    }
}
