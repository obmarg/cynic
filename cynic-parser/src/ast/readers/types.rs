use crate::ast::{ids::TypeId, AstLookup, WrappingType};

use super::{AstId, ReadContext};

#[derive(Clone, Copy)]
pub struct Type<'a>(ReadContext<'a, TypeId>);

impl<'a> Type<'a> {
    pub fn name(&self) -> &'a str {
        self.0.ast.lookup(self.0.ast.lookup(self.0.id).name)
    }

    /// The wrapper types from the outermost to innermost
    pub fn wrappers(&self) -> impl Iterator<Item = crate::ast::WrappingType> + 'a {
        self.0.ast.lookup(self.0.id).wrappers.iter().copied().rev()
    }
}

impl<'a> std::fmt::Display for Type<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let ast = &self.0.ast;

        let crate::ast::Type { name, wrappers } = ast.lookup(self.0.id);

        for wrapping in wrappers.iter().rev() {
            if let WrappingType::List = wrapping {
                write!(f, "[")?;
            }
        }
        write!(f, "{}", ast.lookup(*name))?;
        for wrapping in wrappers.iter() {
            match wrapping {
                WrappingType::NonNull => write!(f, "!")?,
                WrappingType::List => write!(f, "]")?,
            }
        }

        Ok(())
    }
}

impl AstId for TypeId {
    type Reader<'a> = Type<'a>;
}

impl<'a> From<ReadContext<'a, TypeId>> for Type<'a> {
    fn from(value: ReadContext<'a, TypeId>) -> Self {
        Self(value)
    }
}
