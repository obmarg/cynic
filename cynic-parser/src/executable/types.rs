use crate::common::TypeWrappers;

use super::ids::StringId;

pub struct TypeRecord {
    pub name: StringId,
    pub wrappers: TypeWrappers,
}
