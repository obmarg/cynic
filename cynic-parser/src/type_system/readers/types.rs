use crate::{common::WrappingType, type_system::ids::TypeId, AstLookup};

use super::{ReadContext, TypeSystemId};

#[derive(Clone, Copy)]
pub struct Type<'a>(ReadContext<'a, TypeId>);

impl<'a> Type<'a> {
    pub fn name(&self) -> &'a str {
        self.0
            .document
            .lookup(self.0.document.lookup(self.0.id).name)
    }

    /// The wrapper types from the outermost to innermost
    pub fn wrappers(&self) -> impl Iterator<Item = WrappingType> + 'a {
        self.0.document.lookup(self.0.id).wrappers.iter()
    }
}

impl<'a> std::fmt::Display for Type<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let ast = &self.0.document;

        let crate::type_system::Type { name, wrappers } = ast.lookup(self.0.id);

        let wrappers = wrappers.iter().collect::<Vec<_>>();
        for wrapping in &wrappers {
            if let WrappingType::List = wrapping {
                write!(f, "[")?;
            }
        }
        write!(f, "{}", ast.lookup(*name))?;
        for wrapping in wrappers.iter().rev() {
            match wrapping {
                WrappingType::NonNull => write!(f, "!")?,
                WrappingType::List => write!(f, "]")?,
            }
        }

        Ok(())
    }
}

impl TypeSystemId for TypeId {
    type Reader<'a> = Type<'a>;
}

impl<'a> From<ReadContext<'a, TypeId>> for Type<'a> {
    fn from(value: ReadContext<'a, TypeId>) -> Self {
        Self(value)
    }
}
