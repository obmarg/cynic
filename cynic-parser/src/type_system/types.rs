use crate::{
    common::{TypeWrappers, WrappingType},
    type_system::ids::TypeId,
    AstLookup,
};

use super::{ReadContext, StringId, TypeSystemId};

pub struct TypeRecord {
    pub name: StringId,
    pub wrappers: TypeWrappers,
}

#[derive(Clone, Copy)]
pub struct Type<'a>(ReadContext<'a, TypeId>);

impl<'a> Type<'a> {
    pub fn name(&self) -> &'a str {
        self.0
            .document
            .lookup(self.0.document.lookup(self.0.id).name)
    }

    pub fn is_list(&self) -> bool {
        self.wrappers().any(|wrapper| wrapper == WrappingType::List)
    }

    pub fn is_non_null(&self) -> bool {
        self.wrappers().next() == Some(WrappingType::NonNull)
    }

    /// The wrapper types from the outermost to innermost
    pub fn wrappers(&self) -> impl Iterator<Item = WrappingType> + 'a {
        self.0.document.lookup(self.0.id).wrappers.iter()
    }
}

impl std::fmt::Display for Type<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let ast = &self.0.document;

        let TypeRecord { name, wrappers } = ast.lookup(self.0.id);

        let wrappers = wrappers.iter().collect::<Vec<_>>();
        for wrapping in &wrappers {
            if let WrappingType::List = wrapping {
                write!(f, "[")?;
            }
        }
        write!(f, "{}", ast.lookup(*name))?;
        for wrapping in wrappers.iter().rev() {
            match wrapping {
                WrappingType::NonNull => write!(f, "!")?,
                WrappingType::List => write!(f, "]")?,
            }
        }

        Ok(())
    }
}

impl TypeSystemId for TypeId {
    type Reader<'a> = Type<'a>;
}

impl<'a> From<ReadContext<'a, TypeId>> for Type<'a> {
    fn from(value: ReadContext<'a, TypeId>) -> Self {
        Self(value)
    }
}
