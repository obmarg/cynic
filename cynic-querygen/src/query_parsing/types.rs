use inflector::Inflector;
use std::rc::Rc;
use uuid::Uuid;

use super::{normalisation::Variable, value::TypedValue};
use crate::schema::{EnumDetails, InputField, OutputFieldType};
use crate::Error;

// TODO: NOt jazzed about the name of this file as
// graphql has types so it almost sounds like this module is dealing with those...

pub struct Output<'query, 'schema> {
    pub query_fragments: Vec<QueryFragment<'query, 'schema>>,
    pub input_objects: Vec<InputObject<'schema>>,

    // TODO: I want enums sorted by name & unique
    pub enums: Vec<EnumDetails<'schema>>,

    // TODO: Scalars I also want sorted by name & unique.
    pub scalars: Vec<Scalar<'schema>>,

    // TODO: Argument structs can be sorted by name.
    pub argument_structs: Vec<(String, Rc<ArgumentStruct<'query, 'schema>>)>,
}

pub struct Scalar<'schema>(pub &'schema str);

#[derive(Debug, PartialEq)]
pub struct QueryFragment<'query, 'schema> {
    pub fields: Vec<OutputField<'query, 'schema>>,
    pub target_type: String,
    pub argument_struct_name: Option<String>,

    pub name: String,
}

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct ArgumentStruct<'query, 'schema> {
    pub id: Uuid,
    pub(super) name: String,
    pub fields: Vec<ArgumentStructField<'query, 'schema>>,
}

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum ArgumentStructField<'query, 'schema> {
    Variable(Variable<'query, 'schema>),
    NestedStruct(Rc<ArgumentStruct<'query, 'schema>>),
}

impl<'query, 'schema> ArgumentStructField<'query, 'schema> {
    pub fn name(&self) -> String {
        match self {
            ArgumentStructField::Variable(var) => var.name.to_string(),
            ArgumentStructField::NestedStruct(arg_struct) => arg_struct.name.to_snake_case(),
        }
    }

    pub fn type_spec(&self) -> String {
        match self {
            ArgumentStructField::Variable(var) => var.value_type.type_spec().to_string(),
            ArgumentStructField::NestedStruct(arg_struct) => arg_struct.name.clone(),
        }
    }
}

/*
impl<'query, 'schema> ArgumentStruct<'query, 'schema> {
    fn from_variables(
        variables: Vec<Variable<'query, 'schema>>,
        query_name: Option<String>,
    ) -> Option<ArgumentStruct<'query, 'schema>> {
        if variables.is_empty() {
            return None;
        }

        Some(ArgumentStruct {
            name: format!("{}Arguments", query_name.unwrap_or("".to_string())),
            fields: variables,
        })
    }
}
*/

#[derive(Debug, PartialEq)]
pub struct OutputField<'query, 'schema> {
    pub name: &'schema str,
    pub field_type: OutputFieldType<'schema>,

    pub arguments: Vec<FieldArgument<'query, 'schema>>,
}

#[derive(Debug, PartialEq)]
pub struct FieldArgument<'query, 'schema> {
    pub name: &'schema str,
    value: TypedValue<'query, 'schema>,
}

impl<'query, 'schema> FieldArgument<'query, 'schema> {
    pub fn new(name: &'schema str, value: TypedValue<'query, 'schema>) -> Self {
        FieldArgument { name, value }
    }

    pub fn to_literal(&self) -> Result<String, Error> {
        self.value.to_literal(&self.value.value_type())
    }
}

#[derive(Debug, PartialEq)]
pub struct InputObject<'schema> {
    pub name: String,
    pub fields: Vec<InputField<'schema>>,
}
