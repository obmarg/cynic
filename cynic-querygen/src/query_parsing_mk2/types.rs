use graphql_parser::query::{Definition, Document, OperationDefinition, Selection, SelectionSet};
use std::rc::Rc;

use super::{normalisation::OperationVariable, Value};
use crate::schema::{
    self, EnumType, InputField, InputFieldType, InputValue, OutputFieldType, ScalarTypeExt, Type,
    TypeDefinition,
};
use crate::{value_ext::ValueExt, Error, GraphPath, TypeExt, TypeIndex};

pub struct Output<'schema, 'query> {
    pub query_fragments: Vec<QueryFragment<'schema, 'query>>,
    pub input_objects: Vec<InputObject<'schema>>,

    // TODO: I want enums sorted by name & unique
    pub enums: Vec<Enum<'schema>>,

    // TODO: Scalars I also want sorted by name & unique.
    pub scalars: Vec<Scalar<'schema>>,

    // TODO: Argument structs can be sorted by name.
    pub argument_structs: Vec<ArgumentStruct<'schema, 'query>>,
}

pub struct Scalar<'schema>(pub &'schema str);

#[derive(Debug, PartialEq)]
pub struct QueryFragment<'schema, 'query> {
    pub fields: Vec<Field<'schema, 'query>>,
    pub target_type: String,
    pub argument_struct_name: Option<String>,

    pub name: String,
}

impl QueryFragment<'_, '_> {
    fn uses_arguments(&self) -> bool {
        self.fields.iter().any(|f| !f.arguments.is_empty())
    }
}

#[derive(Debug, PartialEq)]
pub struct Enum<'schema> {
    pub def: EnumType<'schema>,
}

#[derive(Debug, PartialEq)]
pub struct ArgumentStruct<'schema, 'query> {
    pub name: String,
    pub fields: Vec<OperationVariable<'query, 'schema>>,
}

impl<'schema, 'query> ArgumentStruct<'schema, 'query> {
    fn from_variables(
        variables: Vec<OperationVariable<'query, 'schema>>,
        query_name: Option<String>,
    ) -> Option<ArgumentStruct<'schema, 'query>> {
        if variables.is_empty() {
            return None;
        }

        Some(ArgumentStruct {
            name: format!("{}Arguments", query_name.unwrap_or("".to_string())),
            fields: variables,
        })
    }
}

#[derive(Debug, PartialEq)]
pub struct Field<'schema, 'query> {
    pub name: &'schema str,
    pub field_type: OutputFieldType<'schema>,

    pub arguments: Vec<FieldArgument<'schema, 'query>>,
}

#[derive(Debug, PartialEq)]
pub struct FieldArgument<'schema, 'query> {
    pub name: &'schema str,
    value: Value<'query>,
    field_type: InputFieldType<'schema>,
}

impl<'schema, 'query> FieldArgument<'schema, 'query> {
    pub fn new(
        name: &'schema str,
        value: Value<'query>,
        field_type: InputFieldType<'schema>,
    ) -> Self {
        FieldArgument {
            name,
            value,
            field_type,
        }
    }

    pub fn to_literal(&self, type_index: &TypeIndex) -> Result<String, Error> {
        self.value.to_literal(&self.field_type)
    }
}

#[derive(Debug, PartialEq)]
pub struct InputObject<'schema> {
    pub name: String,
    pub fields: Vec<InputField<'schema>>,
}
