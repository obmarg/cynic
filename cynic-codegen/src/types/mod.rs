//! This module concerns itself with rust types and how they interact
//! with graphql types.

mod alignment;
mod parsing;
mod validation;

pub use self::{
    alignment::{align_input_type, align_output_type},
    validation::{
        check_input_types_are_compatible, check_spread_type, check_types_are_compatible,
        outer_type_is_option, CheckMode,
    },
};
