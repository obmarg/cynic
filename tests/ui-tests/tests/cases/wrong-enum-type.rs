#![allow(unused_imports)]

fn main() {}

#[cynic::query_module(
    schema_path = r#"./../../../schemas/github.graphql"#,
    query_module = "query_dsl"
)]
mod queries {
    use super::{query_dsl, types::*};

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(graphql_type = "CheckSuite")]
    pub struct CheckSuite {
        // Note: this is the wrong underlying enum type
        // Should be CheckStatusState
        pub status: CheckConclusionState,
        pub conclusion: Option<CheckConclusionState>,
    }

    #[derive(cynic::Enum, Clone, Copy, Debug)]
    #[cynic(graphql_type = "CheckConclusionState")]
    pub enum CheckConclusionState {
        ActionRequired,
        Cancelled,
        Failure,
        Neutral,
        Skipped,
        Stale,
        Success,
        TimedOut,
    }
}

#[cynic::query_module(
    schema_path = r#"./../../../schemas/github.graphql"#,
    query_module = "query_dsl"
)]
mod types {
    #[derive(cynic::Scalar, Debug, Clone)]
    pub struct Date(pub String);

    #[derive(cynic::Scalar, Debug, Clone)]
    pub struct DateTime(pub String);

    #[derive(cynic::Scalar, Debug, Clone)]
    pub struct GitObjectID(pub String);

    #[derive(cynic::Scalar, Debug, Clone)]
    pub struct GitRefname(pub String);

    #[derive(cynic::Scalar, Debug, Clone)]
    pub struct GitSSHRemote(pub String);

    #[derive(cynic::Scalar, Debug, Clone)]
    pub struct GitTimestamp(pub String);

    #[derive(cynic::Scalar, Debug, Clone)]
    pub struct Html(pub String);

    #[derive(cynic::Scalar, Debug, Clone)]
    pub struct PreciseDateTime(pub String);

    #[derive(cynic::Scalar, Debug, Clone)]
    pub struct Uri(pub String);

    #[derive(cynic::Scalar, Debug, Clone)]
    pub struct X509Certificate(pub String);
}

mod query_dsl {
    use super::types::*;
    cynic::query_dsl!(r#"./../../../schemas/github.graphql"#);
}
