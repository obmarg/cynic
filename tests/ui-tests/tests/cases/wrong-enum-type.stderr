error[E0277]: the trait bound `schema::CheckConclusionState: IsFieldType<CheckStatusState>` is not satisfied
  --> tests/cases/wrong-enum-type.rs:13:21
   |
13 |         pub status: CheckConclusionState,
   |                     ^^^^^^^^^^^^^^^^^^^^ the trait `IsFieldType<CheckStatusState>` is not implemented for `schema::CheckConclusionState`
   |
note: required by a bound in `SelectionBuilder::<'a, SchemaType, Variables>::select_field`
  --> $WORKSPACE/cynic/src/queries/builders.rs
   |
   |         FieldType: IsFieldType<SchemaType::Type>,
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `SelectionBuilder::<'a, SchemaType, Variables>::select_field`
